x-pack/plugin/upgrade/src/main/java/org/elasticsearch/xpack/upgrade/SecurityIndexUpgradeCheck.java

```diff
@@ -0,0 +1,245 @@
+                listener.onFailure(e);
+            }, e1 -> {
+                e.addSuppressed(e1);
+                listener.onFailure(e);
+            }));
+        // check all nodes can handle the new index format
+        checkMasterAndDataNodeVersion(state);
+        // create the two new indices, but without their corresponding aliases
+        createNewSecurityIndices(parentAwareClient, ActionListener.wrap(createIndicesResponse -> {
+            // set a read-only block on the original .security-6 index
+            setReadOnlyBlock(INTERNAL_SECURITY_INDEX, ActionListener.wrap(setReadOnlyResponse -> {
+                // firstly, reindex all the tokens to the new tokens index
+                reindex(parentAwareClient, INTERNAL_SECURITY_INDEX, INTERNAL_SECURITY_TOKENS_INDEX, tokensQuery,
+                        ActionListener.wrap(reindexTokensResponse -> {
+                            // create the alias pointing to the new tokens index; now the new tokens index is writable anew
+                            parentAwareClient.admin().indices().prepareAliases()
+                                    .addAlias(INTERNAL_SECURITY_TOKENS_INDEX, SECURITY_TOKENS_ALIAS_NAME)
+                                    .execute(ActionListener.wrap(tokensAliasResponse -> {
+                                        // now reindex all the other non-token security objects to the new .security-7 index
+                                        reindex(parentAwareClient, INTERNAL_SECURITY_INDEX, NEW_INTERNAL_SECURITY_INDEX, nonTokensQuery,
+                                                ActionListener.wrap(reindexNonTokensResponse -> {
+                                                    // move the .security alias to the new index, and remove the old .security-6 index
+                                                    parentAwareClient.admin().indices().prepareAliases()
+                                                            .addAlias(NEW_INTERNAL_SECURITY_INDEX, SECURITY_ALIAS_NAME)
+                                                            .removeIndex(INTERNAL_SECURITY_INDEX)
+                                                            .execute(ActionListener.wrap(aliasResponse -> {
+                                                                // merge both reindex responses to return to the action listener
+                                                                listener.onResponse(new BulkByScrollResponse(
+                                                                        Arrays.asList(reindexTokensResponse, reindexNonTokensResponse), null));
+                                                            }, removeReadOnlyBlock));
+                                                }, removeReadOnlyBlock));
+                                    }, removeReadOnlyBlock));
+                        }, removeReadOnlyBlock));
+            }, listener::onFailure));
+        }, listener::onFailure));
```

Comment:
I think the upgrade method is a good place to deep dive in the PR review.

Reply:
This does not have tests, I know, I'll add them after a few review rounds.

Reply:
I think we also should go ahead and remove any documents with a type of `invalidated-token`. These documents are really not relevant anymore

Reply:
Also this isn't available yet, but this might be a good place to make use of the `StepListener` from #37327